drivers/smb/driver.go

```diff
@@ -0,0 +1,157 @@
+package smb
+
+import (
+	"context"
+	"errors"
+	"path/filepath"
+
+	"github.com/alist-org/alist/v3/internal/driver"
+	"github.com/alist-org/alist/v3/internal/model"
+	"github.com/alist-org/alist/v3/pkg/utils"
+
+	"github.com/hirochachacha/go-smb2"
+)
+
+type SMB struct {
+	model.Storage
+	Addition
+	fs *smb2.Share
+}
+
+func (d *SMB) Config() driver.Config {
+	return config
+}
+
+func (d *SMB) GetAddition() driver.Additional {
+	return d.Addition
+}
+
+func (d *SMB) Init(ctx context.Context, storage model.Storage) error {
+	d.Storage = storage
+	err := utils.Json.UnmarshalFromString(d.Storage.Addition, &d.Addition)
+	if err != nil {
+		return err
+	}
+	return d.initFS()
+}
+
+func (d *SMB) Drop(ctx context.Context) error {
+	if d.fs != nil {
+		_ = d.fs.Umount()
+	}
+	return nil
+}
+
+func (d *SMB) List(ctx context.Context, dir model.Obj, args model.ListArgs) ([]model.Obj, error) {
+	fullPath := d.getSMBPath(dir)
+	rawFiles, err := d.fs.ReadDir(fullPath)
+	if err != nil {
+		return nil, err
+	}
+	var files []model.Obj
+	for _, f := range rawFiles {
+		file := model.ObjThumb{
+			Object: model.Object{
+				Name:     f.Name(),
+				Modified: f.ModTime(),
+				Size:     f.Size(),
+				IsFolder: f.IsDir(),
+			},
+		}
+		files = append(files, &file)
+	}
+	return files, nil
+}
+
+//func (d *SMB) Get(ctx context.Context, path string) (model.Obj, error) {
+//	// this is optional
+//	return nil, errs.NotImplement
+//}
+
+func (d *SMB) Link(ctx context.Context, file model.Obj, args model.LinkArgs) (*model.Link, error) {
+	fullPath := file.GetPath()
+	return &model.Link{FilePath: &fullPath}, nil
```

Comment:
Is it correct to return the local file path here?

Reply:
Yes, I have tried to return the path with the prefix `.` and it gives a 500 error, which takes me some time to find out.

Reply:
But wouldn't that just return the native file only? 

it is possible that smb on another machine.

Reply:
My bad, I think I do not understand the meaning of the function, trying to fix it now.

Reply:
I think it works now for smb on another machine.

Reply:
Ok. This is function return one of three:
- path of native file
- download url of web file
- reader closer

For this, It maybe like:
```go
f,_ := d.fs.Open(converted_path)
return &Link{
    Data: f
}
```
In a rough sense